### PR TITLE
Generate a valid URL for link notification

### LINKED ISSUE #23197: Link to the folder/file is broken/wrong in the email sent via internal share.
### Steps to reproduce
As a logged-in user,

1. Try to share a folder with another ownCloud user(i.e., via internal share).
2. Notify that user via email by selecting that check-box.

### Expected behaviour
The ownCloud user should receive a mail such as the one in the screenshot. The email contains a link to the shared folder/file. On clicking that link, we should be redirected to the respective folder/file in the ownCloud app.

### Actual behaviour
The link is broken/wrong. The screenshot is attached.

### Server configuration
**Operating system**: Ubuntu 14.04

**Web server:** Apache/2.4.7

**Database:** mysql  Ver 14.14 Distrib 5.6.28, for debian-linux-gnu (x86_64)

**PHP version:** PHP 5.5.9-1ubuntu4.14

**ownCloud version:** 9.1.0 pre alpha

**Where did you install ownCloud from:** Git

**Signing status (ownCloud 9.0 and above):**

```
Integrity checker has been disabled. Integrity cannot be verified.
```

**List of activated apps:**

```
Enabled:
  - comments: 0.3.0
  - dav: 0.2.1
  - federatedfilesharing: 0.2.0
  - federation: 0.1.0
  - files: 1.5.1
  - files_sharing: 0.10.0
  - files_trashbin: 0.9.0
  - files_versions: 1.3.0
  - gallery: 15.0.0
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
  - updatenotification: 0.2.0
Disabled:
  - encryption
  - files_external
  - testing
  - user_ldap
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "oc4z46udfiog",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost:8080"
        ],
        "datadirectory": "\/var\/www\/GSoC\/data",
        "overwrite.cli.url": "http:\/\/localhost:8080",
        "dbtype": "mysql",
        "version": "9.1.0.0",
        "dbname": "ownCloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "mail_from_address": "jalpreet.cse",
        "mail_smtpmode": "smtp",
        "mail_domain": "gmail.com",
        "mail_smtpsecure": "ssl",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "smtp.gmail.com",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "loglevel": 0
    }
}

```

**Are you using external storage, if yes which one:** No

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Chrome 49.0.2623.75 (64-bit)

**Operating system:** Ubuntu 14.04

#### ownCloud log (data/owncloud.log)
```
[Log](https://gist.github.com/imjalpreet/2188043ccc6bcfcc487f)

```
issuecore2
issuecore

### PR DESCRIPTION
**Pull request by rullzer**
fixes #23197

@imjalpreet can you verify this fix (as I don't run apache etc).
- [x] Update unit tests to actually check this behaviour

### CODE REVIEW

**Comment by imjalpreet** [comment]
@rullzer I tested it. It is working now. Though we need to still correct the link. The link which is being sent in the mail right now is not correct. I will give you an example:
Suppose we have the following directory structure: 
**Test/Test2** and we share **Test2**(both of them are folders). Now in the link that is generated now we are just sending the name of the folder rather than the path, i.e., we have _link?dir=Test2_, which is wrong and that's why it always goes to the home directory of the files app(**Test** in this case). 
According to me, the mistake is in the following line because we are sending directory name in the arguments instead of the path ( [Here](https://github.com/owncloud/core/blob/master/lib/private/share/mailnotifications.php#L120) ).

File: core/ajax/share.php [hunk 1]
```diff
@@ -70,7 +70,8 @@
 				\OC::$server->getL10N('lib'),
 				\OC::$server->getMailer(),
 				\OC::$server->getLogger(),
-				$defaults
+				$defaults,
+				\OC::$server->getURLGenerator(),
```
Comment (imjalpreet) [inline core/ajax/share.php]: @rullzer There is a syntax error due to the extra comma in this line.

Comment (rullzer) [inline core/ajax/share.php]: Ah yes... commited to fast. let me force push

**Comment by rullzer** [comment]
@imjalpreet no that is correct.

Because if `userA` shares `Test/Test2` to `userB`. `userB` will just see `Test2` appear in his root folder.

**Comment by imjalpreet** [comment]
@rullzer Ah Yes, I agree with that.
Sorry, I was doing one mistake. Actually, I had made two accounts in my ownCloud app to test but I forgot to logout from the one I sent the email from. So, that's why the link that I was trying to open was always redirecting to the home of the files app. Basically, I had to login to the other account. (Do you get what I am trying to say?)
But due to this I came across another issue(maybe an improvement). I think when I click on the link from my email and it redirects to the owncloud app, then it should ask me to login to my account in which I got the share link instead of trying to go the link in the current logged in account. If it doesn't ask me that, then if in my currently logged in account if I have the directory with the same name it will go to that, which I think is not the best way. Instead, we can think of improving this by asking the user to login to the account with which the link has been shared.

@rullzer What do you think?

**Comment by imjalpreet** [comment]
Otherwise, I have tested this. It's :+1:

**Comment by rullzer** [comment]
@imjalpreet well that is a very specific use case. And the link currently does not contain that info. I think it will be more work to implement than users using it. I mean typically you only have 1 account on an instance.

Thanks for testing. I'll pin it down with extending the unit test then!

**Comment by imjalpreet** [comment]
> @imjalpreet well that is a very specific use case. And the link currently does not contain that info. I think it will be more work to implement than users using it. I mean typically you only have 1 account on an instance.

Yeah, I agree. We can implement this in the future maybe. 

> Thanks for testing. I'll pin it down with extending the unit test then!

:+1:

**Comment by rullzer** [comment]
@karlitschek we should backport this to stable9

**Comment by rullzer** [comment]
Review time: @PVince81 @LukasReschke @MorrisJobke @nickvergessen

**Comment by LukasReschke** [comment]
:+1:

**Comment by karlitschek** [comment]
please backport :+1:

**Comment by PVince81** [comment]
:+1:

@rullzer did it work correctly in OC <= 8.2 ?

**Comment by rullzer** [comment]
@PVince81 I think so. But I'll double check...

**Comment by rullzer** [comment]
@PVince81 no 8.2 is fine

**Pull request merged**